src/main/java/de/medizininformatik_initiative/processes/common/fhir/client/token/OAuth2TokenClient.java

```diff
@@ -0,0 +1,231 @@
+package de.medizininformatik_initiative.processes.common.fhir.client.token;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.InetSocketAddress;
+import java.net.ProxySelector;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.file.Path;
+import java.security.KeyManagementException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+import java.time.Duration;
+import java.util.Base64;
+
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManagerFactory;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.InitializingBean;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import de.rwh.utils.crypto.io.CertificateReader;
+
+public class OAuth2TokenClient implements TokenClient, InitializingBean
+{
+	private static final Logger logger = LoggerFactory.getLogger(OAuth2TokenClient.class);
+
+	private static final String TLS = "TLS";
+
+	private static final String HEADER_CONTENT_TYPE = "Content-Type";
+	private static final String MIME_TYPE_FORM_URLENCODED = "application/x-www-form-urlencoded";
+	private static final String HEADER_AUTHORIZATION = "Authorization";
+	private static final String HEADER_AUTHORIZATION_PROXY = "Proxy-Authorization";
+	private static final String PREFIX_BASIC_AUTH = "Basic ";
+	private static final String GRANT_TYPE_CLIENT_CREDENTIALS = "grant_type=client_credentials";
+
+	private final String issuerUrl;
+	private final String clientId;
+	private final String clientSecret;
+
+	private final int connectTimeout;
+	private final int socketTimeout;
+
+	private final Path trustStorePath;
+
+	private final String proxyUrl;
+	private final String proxyUsername;
+	private final String proxyPassword;
+
+	private final ObjectMapper objectMapper;
+
+	static
+	{
+		System.setProperty("jdk.http.auth.tunneling.disabledSchemes", "");
+	}
+
+	public OAuth2TokenClient(String issuerUrl, String clientId, String clientSecret, int connectTimeout,
+			int socketTimeout, Path trustStorePath, String proxyUrl, String proxyUsername, String proxyPassword)
+	{
+		this(issuerUrl, clientId, clientSecret, connectTimeout, socketTimeout, trustStorePath, proxyUrl, proxyUsername,
+				proxyPassword, new ObjectMapper());
+	}
+
+	public OAuth2TokenClient(String issuerUrl, String clientId, String clientSecret, int connectTimeout,
+			int socketTimeout, Path trustStorePath, String proxyUrl, String proxyUsername, String proxyPassword,
```

Comment:
```suggestion
			int socketTimeout, Path trustStorePath, String proxyUrl, String proxyUsername, char[] proxyPassword,
```

Reply:
See comment above for preemptive proxy authorization.